samples/dotnet/openapi-skills/Program.cs

```diff
@@ -0,0 +1,227 @@
+            string planResult = await PlanGitHubSkill(gitHubOptions, planner, chatHistory, input, planResultTokenAllowance);
+            if (!string.IsNullOrWhiteSpace(planResult))
+            {
+                chatHistory.AddUserMessage(planResult);
+            }
+
+            // Add the user's input to the chat history.
+            chatHistory.AddUserMessage(input);
+
+            // Remove earlier messages until we are back within our token limit.
+            // (Note this sample does not implement long-term memory)
+            int tokenCount = GPT3Tokenizer.Encode(JsonSerializer.Serialize(chatHistory)).Count;
+            while (tokenCount > aiOptions.TokenLimit)
+            {
+                chatHistory.Messages.RemoveAt(0);
+                tokenCount = GPT3Tokenizer.Encode(JsonSerializer.Serialize(chatHistory)).Count;
+            }
+            Console.WriteLine($"(tokens: {tokenCount})");
+
+            // Send the chat history to the AI for a response.
+            string reply = await chatGPT.GenerateMessageAsync(chatHistory);
+            chatHistory.AddAssistantMessage(reply);
+
+            Console.WriteLine("----------------");
+            Console.WriteLine($"Assistant: {reply}");
+        }
+    }
+
+    /// <summary>
+    /// Run the planner to decide whether to run the GitHub skill function and add the result to the chat history.
+    /// </summary>
+    private static async Task<string> PlanGitHubSkill(GitHubSkillOptions gitHubOptions, ActionPlanner planner, OpenAIChatHistory chatHistory, string input, int tokenAllowance)
+    {
+        // Ask the planner to create a plan based off the user's input. If the plan elicits no steps, continue normally.
+        Plan plan = await planner.CreatePlanAsync(input);
+        if (plan.Steps.Count > 0)
+        {
+            // Make sure the plan's state has the GitHub skill configuration values.
+            plan.State.Set("repo", gitHubOptions.Repository);
+            plan.State.Set("owner", gitHubOptions.Owner);
+
+            // Run the plan
+            SKContext planContext = await plan.InvokeAsync(logger: logger);
+            if (planContext.ErrorOccurred)
+            {
+                logger.LogError("{0}", planContext.LastErrorDescription);
+            }
+
+            if (!TryExtractJsonFromOpenApiPlanResult(planContext.Result, out string planResult))
+            {
+                planResult = planContext.Result;
+            }
+
+            // GitHub responses can be very lengthy - optimize the output so we don't immediately go beyond token limits.
+            planResult = OptimizeGitHubPullRequestResponse(planResult, tokenAllowance);
+
+            return planResult;
+        }
+        return string.Empty;
+    }
+
+    /// <summary>
+    /// Reduce the size of GitHub PullRequest responses.
+    /// </summary>
+    private static string OptimizeGitHubPullRequestResponse(string planResult, int tokenAllowance)
```

Comment:
I'm curious if you've tried making these SKFunctions and allowed SequentialPlanner to connect the dots.

Reply:
I have not, but I'll leave that as an exercise for another day since the focus of this example is the OpenAPI skill.